### PR TITLE
image manifest: support use as artifact

### PR DESCRIPTION
**Pull request by AaronFriel**
Implementations lack any information to determine whether the `config` property of an Image Manifest describes the `artifactType` of the manifest, or the media type of the referenced content. This is necessitated by the removal of artifact manifest (#999) and addition of scratch descriptors (#1023) to support artifacts.

If #999 is withdrawn, this pull request should be withdrawn as well and #1028 considered instead.

### CODE REVIEW

File: considerations.md
Comment (sudo-bmitch) [inline considerations.md:3]: "MUST NOT generate error if they encounter an unknown manifest" implies the `MANIFEST_UNKNOWN` error code in distribution spec is no longer valid. I don't think we can do that since it's not how any registry I'm aware of works today, everyone validates the manifest media type.

I would like to say "Implementations storing or copying content MUST NOT modify or alter the content in a way that would change the digest of the content."

File: descriptor.md
Comment (sudo-bmitch) [inline descriptor.md:21]: I don't think we need this with the addition of `artifactType` and the scratch media type.

**Review by sudo-bmitch — CHANGES_REQUESTED** [review]

File: considerations.md
Comment (AaronFriel) [reply considerations.md:3]: 🙏 thank you for the feedback, @sudo-bmitch. Will #1043 merge today or early next week so I can rebase and incorporate these changes?

Comment (sudo-bmitch) [reply considerations.md:3]: That's a question for the other maintainers. We can merge once we get two approvals.

File: descriptor.md
Comment (AaronFriel) [reply descriptor.md:21]: Removed

File: considerations.md
Comment (AaronFriel) [reply considerations.md:3]: I've pulled this out into a separate PR, replacing the line with yours

File: manifest.md
Comment (sudo-bmitch) [inline manifest.md:34]: This makes sense for storing and distributing artifacts. But for tooling that processes an artifact, they may reject unknown types. E.g. if a tool parses an SBOM in json formats, and the artifactType is for an xml format, that tool should be allowed to generate an error. Or helm told to deploy something that is not a helm chart.

File: manifest.md
Comment (sudo-bmitch) [inline manifest.md:90]: Image runtimes should be allowed to generate an error if they cannot process the layers. Same for artifact processing tools that encounter an unexpected layer (e.g. helm trying to process a gif layer).

**Review by sudo-bmitch — CHANGES_REQUESTED** [review]
The wording covers storage and distribution of content, but we need to include content producers and consumers (like image runtimes) in this spec.

File: manifest.md
Comment (AaronFriel) [reply manifest.md:90]: How would you phrase that distinction?

Does it make sense for us to define consumers (runtimes) vs distributors (registries, proxies) somewhere?

Comment (sudo-bmitch) [reply manifest.md:90]: We say "implementations storing or copying content" above. That phrasing could work here too.

**Comment by sudo-bmitch** [comment]
I'd want to hear from registry operators about this before adding my own review. I could understand some of them concerned that they must accept unknown content and the only blocks that are being permitted are deny lists rather than allow lists.

**Comment by sajayantony** [comment]
Writing up my thoughts here since I have been mulling on this for quite some time and feel I owe a response to @AaronFriel who has been on multiple calls.

We do have customers who do want to limit the types that the registry should accept. This doesn't mean much since you can masquerade anything as an image. Implementations have already shown that.

The main issue I'm reading here is that there must be a way for artifact authors to get a sense that OCI registries are the way to store artifacts and not worry that their type would be rejected and my feeling is that it's a distribution spec issue as the image-manifest already calls out how an artifact may be represented.  

I'm not sure if rewording of 'MUST be ignored' to 'MUST not generate an error' would change any implementation behaviors, but https://github.com/opencontainers/image-spec/pull/1030/files#diff-8f83bd1bc119af1add2e038b7c3aeade8b3845efed06a6971d1ee8740d7a3325R150 does add the fix for OCI layout which can be used for artifacts as well.

File: config.md
Comment (vbatts) [inline config.md:251]: 👍 This has been a confusing line

File: considerations.md
Comment (vbatts) [inline considerations.md:3]: While I agree with this sentence regarding altering of content, it could be straying away from the topic whether to generate errors or not

**Review by vbatts — APPROVED** [review]
I'm on board with these changes. This wording has been confusing and needed fixing for a while.

**Comment by vbatts** [comment]
@sajayantony interesting. It's like the customers (and their tools) would need an endpoint or HTTP headers to discover what types are accepted. If it can vary per customer or even per registry repo ...

**Review by sajayantony — APPROVED** [review]
If this enables sentiment of artifact types as first class +1

File: considerations.md
Comment (tianon) [reply considerations.md:3]: Agreed, this feels like a distribution-spec concern, not an image-spec concern -- a registry that intentionally modifies content pushed to it has some interesting potential use cases, and wouldn't be compliant to the distribution spec, but absolutely could be image-spec complaint, and that would be OK, I think?

**Pull request merged**
